libmamba/src/solv-cpp/pool.hpp

```diff
@@ -0,0 +1,214 @@
+        auto get_dependency_name(DependencyId id) const -> std::string_view;
+        auto get_dependency_version(DependencyId id) const -> std::string_view;
+        auto get_dependency_relation(DependencyId id) const -> std::string_view;
+        auto dependency_to_string(DependencyId id) const -> std::string;
+
+        void create_whatprovides();
+        template <typename UnaryFunc>
+        void for_each_whatprovides_id(DependencyId dep, UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_whatprovides(DependencyId dep, UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_whatprovides(DependencyId dep, UnaryFunc func);
+
+        auto select_solvables(const ObjQueue& job) const -> ObjQueue;
+
+        auto add_repo(std::string_view name) -> std::pair<RepoId, ObjRepoView>;
+        auto has_repo(RepoId id) const -> bool;
+        auto get_repo(RepoId id) -> std::optional<ObjRepoView>;
+        auto get_repo(RepoId id) const -> std::optional<ObjRepoViewConst>;
+        auto n_repos() const -> std::size_t;
+        auto remove_repo(RepoId id, bool reuse_ids) -> bool;
+        template <typename UnaryFunc>
+        void for_each_repo_id(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_repo(UnaryFunc func);
+        template <typename UnaryFunc>
+        void for_each_repo(UnaryFunc func) const;
+        auto installed_repo() const -> std::optional<ObjRepoViewConst>;
+        auto installed_repo() -> std::optional<ObjRepoView>;
+        void set_installed_repo(RepoId id);
+
+        auto get_solvable(SolvableId id) const -> std::optional<ObjSolvableViewConst>;
+        auto get_solvable(SolvableId id) -> std::optional<ObjSolvableView>;
+        template <typename UnaryFunc>
+        void for_each_solvable_id(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_solvable(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_solvable(UnaryFunc func);
+
+        template <typename Func>
+        void set_debug_callback(Func callback);
+
+    private:
+
+        struct PoolDeleter
+        {
+            void operator()(::Pool* ptr);
+        };
+
+        std::unique_ptr<void, void (*)(void*)> m_user_debug_callback;
+        // Must be deleted before the debug callback
+        std::unique_ptr<::Pool, ObjPool::PoolDeleter> m_pool = nullptr;
+    };
+}
+
+/*******************************
+ *  Implementation of ObjPool  *
+ *******************************/
+
+#include <solv/pool.h>
+
+namespace mamba::solv
+{
+
+    template <typename UnaryFunc>
+    void ObjPool::for_each_repo_id(UnaryFunc func) const
+    {
+        const ::Pool* const pool = raw();
+        const ::Repo* repo = nullptr;
+        RepoId repo_id = 0;
+        FOR_REPOS(repo_id, repo)
+        {
+            func(repo_id);
```

Comment:
`func` can potentially throw. Is it ok here?
If not, consider using `safe_invoke` to wrap the call and gather the errors (see https://github.com/mamba-org/mamba/blob/main/libmamba/include/mamba/core/invoke.hpp) - which lacks documentation, sorry!

Reply:
Throwing is not a problem here.